src/samplers/cpu/cpi/mod.rs

```diff
@@ -91,17 +179,87 @@ impl Sampler for Cpi {
 
         let elapsed = (now - self.prev).as_secs_f64();
 
-        match self.group.read() {
-            Ok(counts) => {
-                for (metric, pc) in &mut self.counters {
-                    metric.set(now, elapsed, counts[&pc]);
+        let mut total_cycles = 0;
+        let mut total_instructions = 0;
+        let mut base_frequency = 1;
+        let mut cur_group_data = vec![];
+        for group in &mut self.groups {
+            match group.read_group() {
+                Ok(counts) => {
+                    let running_us;
+                    match (counts.time_enabled(), counts.time_running()) {
+                        (Some(enable_time), Some(running_time)) => {
+                            if enable_time.as_nanos() != running_time.as_nanos() {
+                                error!("The perf group {} is not always running", group.cpu);
+                                continue;
+                            }
+                            running_us = running_time.as_micros()
+                        }
+                        (_, _) => {
+                            error!(
+                                "The perf group {} has no enabled time and running time",
+                                group.cpu
+                            );
+                            continue;
+                        }
+                    }
+                    if counts.time_enabled() != counts.time_running() {}
+                    let cycles = counts[&(group.cycles)];
+                    let instructions = counts[&(group.instructions)];
+
+                    total_cycles += cycles;
+                    total_instructions += instructions;
+                    let tsc = counts[&(group.tsc)];
+                    base_frequency = tsc / running_us as u64;
+                    BASEFREQUENCY.set(base_frequency as i64);
```

Comment:
@brayniac Maybe we should move the basefrequency to the hwinfo() sampler since it's never changed.

Reply:
That's a good idea. We should probably calculate it once and store it somewhere. It's probably okay to refactor this later.